### ISSUE TITLE
report a warning for returning a value within a forEach() callback

### ISSUE THREAD

**Issue by DartBot** [dart-lang/linter#2138]
_This issue was originally filed by bslesinsk...&#064;gmail.com_

---

The following code is wrong, but looks right. No warning is reported at compile time and no error happens at runtime:

String findFirstError(Map m) {
&nbsp;&nbsp;m.forEach((k,v) {
&nbsp;&nbsp;&nbsp;&nbsp;if (!validate(k,v)) {
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;return &quot;found a problem with ${k}&quot;;
&nbsp;&nbsp;&nbsp;&nbsp;}
&nbsp;&nbsp;});
&nbsp;&nbsp;return &quot;ok&quot;;
}

By analogy with regular for loops this looks like it should return early when it finds the first error, but actually the return value from the callback is ignored, so it always returns &quot;ok&quot;.

It seems like we should be able to infer that the callback's return type is void, and therefore any return statements within the callback should not return a value.

**Comment by srawlins** [comment]
I think this is a great request for the linter. Accidental return for new users.

**Comment by eernstg** [comment]
This is actually a tricky case, because the inferred return type of a function literal will be based on what is actually returned in the case where this is a subtype of the return type which is required by the context:

```dart
void main() {
  void f(void Function() g) => print(g.runtimeType);
  f(() => 42); // Prints '() => int'.
}
```

So we won't force the function literal `() => 42` to have type `void Function()` just because that's what `f` expects, we will allow it to have its "natural" type `int Function()`.

It is not an error to pass a value of type `int Function()` to a parameter of type `void Function()`
(indeed, `int <: void`, and there's nothing unsound in ignoring an integer), but this means that the error that we'd normally get when "returning a value to void" disappears:

```dart
void h() {
  return 42; // Error, returning an `int` with return type `void`.
}
```

So this is actually an example of a case where we need "voidness preservation", which is the extension of the checks concerned with the type `void` to cases where `void` is just part of the type (e.g., `void Function()` vs. `int Function()`, or `List<void>` vs. `List<int>`).

Voidness preservation as a lint is requested in https://github.com/dart-lang/linter/issues/1124, and it has previously been discussed (e.g., https://github.com/dart-lang/sdk/issues/30352, https://github.com/dart-lang/sdk/issues/34455).

So this issue provides an example which is highly motivating, but other than that it is actually a duplicate of #1124. I won't close it because it could be argued that this special case should be treated separately, but I think it's useful to at least be aware of this connection.
